ens-improvement-proposals/ensip-16-offchain-metadata.md

```diff
@@ -0,0 +1,167 @@
+
+### Implementation
+
+```solidity
+
+interface OffChainResolver {
+    /** @dev Returns the owner of the resolver on L2
+     * @param node
+     * @return owner in bytes32 instead of address to cater for non EVM based owner information
+     */
+    owner(bytes32 node) returns (bytes32 owner);
+
+    isApprovedForAll(address account, address operator) returns (boolean);
+    /** @dev Returns the owner of the resolver on L2
+     * @return name can be l2 chain name or url if offchain
+     * @return coinType according to https://github.com/ensdomains/address-encoder
+     * @return graphqlUrl url of graphql endpoint that provides additional information about the offchain name and its subdomains
+     * @return storageType 0 = EVM, 1 = Non blockchain, 2 = Starknet
+     * @storageLocation = l2 contract address
+     * @return context = an arbitrary bytes string to define the namespace to which a record belongs such as the name owner.
+     */
+    function metadata(bytes calldata name)
+        external
+        view
+        returns (string memory, uint256, string memory, uint8, bytes memory, bytes memory)
+    {
+        return (name, coinType, graphqlUrl, storageType, storageLocation, context);
+    }
+
+    // Optional. If context is dynamic, the event won't be emitted.
+    event MetadataChanged(
+        string name,
+        uint256 coinType,
+        string graphqlUrl,
+        uint8 storageType,
+        bytes storageLocation,
+        bytes context
+    );
+}
+```
+
+```javascript
+const node = namehash('ccipreadsub.example.eth')
+const resolver = await ens.resolver(node)
+const owner = await resolver.owner(node)
+// 0x123...
+const dataLocation = await.resolver.graphqlUrl()
+// {
+//   url: 'http://example.com/ens/graphql',
+// }
+```
+
+##### GgraphQL schema
+
+##### L1
+
+```graphql
+
+type Domain @entity{
+  id
+  metadata: Metadata
+}
+
+type Metadata @entity {
+  "l1 resolver address"
+  id: ID!
+  "Name of the Chain"
+  name: String
+  "coin type"
+  coinType: BigInt
+  "url of the graphql endpoint"
+  graphqlUrl: String
+  "0 for evm, 1 for non blockchain, 2 for starknet"
+  storageType: Int
+  "l2 contract address"
+  storageLocation: Bytes
+  "optional field to store an arbitrary bytes string to define the namespace to which a record belongs"
+  context: Bytes
+  "optional field if the name has expirty date offchain"
+  expiryDate: BigInt
+}
+
+type Resolver @entity {
+  offchain: Offchain
+}
+```
+
+##### L2
+
+```graphql
+type Domain {
+  id: ID! # concatenation of context and namehash delimited by `-`
+  context: Bytes
+  name: String
+  namehash: Bytes
+  labelName: String
+  labelhash: Bytes
+  resolvedAddress: Bytes
+  parent: Domain
+  subdomains: [Domain]
```

Comment:
Can we also add a `subdomainCount` here too?

Also, is pagination (or lack thereof) left up to the implementer to support or not?